dashboard/src/components/PrivateRoute/PrivateRoute.tsx

```diff
@@ -1,26 +1,70 @@
 import * as React from "react";
+import * as Modal from "react-modal";
 import { Redirect, Route, RouteComponentProps, RouteProps } from "react-router";
+import { Auth } from "shared/Auth";
 
 type IRouteComponentPropsAndRouteProps = RouteProps & RouteComponentProps<any>;
 
 interface IPrivateRouteProps extends IRouteComponentPropsAndRouteProps {
   authenticated: boolean;
 }
 
-class PrivateRoute extends React.Component<IPrivateRouteProps> {
+interface IPrivateRouteState {
+  previouslyAuthenticated: boolean;
+}
+
+class PrivateRoute extends React.Component<IPrivateRouteProps, IPrivateRouteState> {
+  public state: IPrivateRouteState = {
+    previouslyAuthenticated: false,
+  };
+
   public render() {
     const { authenticated, component: Component, ...rest } = this.props;
     return <Route {...rest} render={this.renderRouteIfAuthenticated} />;
   }
 
   public renderRouteIfAuthenticated = (props: RouteComponentProps<any>) => {
     const { authenticated, component: Component } = this.props;
-    return authenticated && Component ? (
-      <Component {...props} />
-    ) : (
-      <Redirect to={{ pathname: "/login", state: { from: props.location } }} />
-    );
+    if (authenticated && Component) {
+      if (!this.state.previouslyAuthenticated) {
+        this.setState({ previouslyAuthenticated: true });
+      }
+      return <Component {...props} />;
+    }
+    if (!authenticated && this.state.previouslyAuthenticated && Auth.usingOIDCToken()) {
+      return (
+        <Modal
+          style={{
+            content: {
+              bottom: "auto",
+              left: "50%",
+              marginRight: "-50%",
+              right: "auto",
+              top: "50%",
+              transform: "translate(-50%, -50%)",
+            },
+          }}
+          isOpen={true}
+        >
+          <div>
+            <div className="margin-b-normal">
+              Your session has expired, please reload to refresh your credentials
+            </div>
+            <div className="flex text-c">
+              <button className="button" onClick={this.reload}>
```

Comment:
something interesting I found is that with this modal, when clicking in reload, the actual page gets reloaded while if I just do:

```js
    if (!authenticated && this.state.previouslyAuthenticated && Auth.usingOIDCToken()) {
      this.reload()
    }
```

it gets redirected to the root path. I am not sure the reason of this behavior. Any ideas?

Reply:
We render multiple PrivateRoutes per private route (https://github.com/kubeapps/kubeapps/blob/f4ee15254e0d96fd4ab747a01ad783618da81edb/dashboard/src/containers/RoutesContainer/Routes.tsx#L25) so we might be rendering this modal multiple times. Can you check the DOM to see if that's the case? If so, we need to find another way to do this.

Reply:
Nevermind, I was wrong, the `Switch` ensures we only load one https://github.com/kubeapps/kubeapps/blob/f4ee15254e0d96fd4ab747a01ad783618da81edb/dashboard/src/containers/RoutesContainer/Routes.tsx#L55.